content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -994,7 +996,68 @@ Here is the declarations type mapping between OpenAPIv3 and CEL type:
+
+
+#### Transition Rules
 
+A rule that contains an expression referencing the identifier `oldSelf` is implicitly considered a
+"transition rule". Transition rules allow schema authors to prevent certain transitions between two
+otherwise valid states. For example:
+
+```yaml
+type: string
+enum: ["low", "medium", "high"]
+x-kubernetes-validations:
+- rule: "!(self == 'high' && oldSelf == 'low') && !(self == 'low' && oldSelf == 'high')"
+  message: cannot transition directly between 'low' and 'high'
+```
+
+Errors will be generated on CRD writes if a schema node contains a transition rule that can never be
+applied, e.g. "*path*: update rule *rule* cannot be set on schema because the schema or its parent
+schema is not mergeable".
+
+##### Use Cases
+
+| Use Case                                                          | Rule
+| --------                                                          | --------
+| Immutability                                                      | `self.foo == oldSelf.foo`
+| Prevent modification/removal once assigned                        | `oldSelf != 'bar' \|\| self == 'bar'` or `!has(oldSelf.field) \|\| has(self.field)`
+| Append-only set                                                   | `self.all(element, element in oldSelf)`
+| If previous value was X, new value can only be A or B, not Y or Z | `oldSelf != 'X' \|\| self in ['A', 'B']`
+| Nondecreasing counters                                            | `self >= oldSelf`
+
+#### Resource Constraints
+
+CEL expressions have the potential to consume unacceptable amounts of API server resources. We constrain the resource utilization in following ways:
+- Validation of CEL expression's "cost" when a CEL expression is written to a field in a CRD (at CRD creation/update time)
+- Runtime cost budget during CEL evaluation
+  - CEL validation might fail due to runtime cost budget exceed with error message `validation failed due to running out of cost budget, no further validation rules will be run`
+  - CEL validation might fail due to cost limit exceed per expression with message `operation cancelled: actual cost limit exceeded: no further validation rules will be run due to call cost exceeds limit for rule:{$rule}`
+- Go context cancellation to bound CEL expression evaluation to the request lifetime
+
+Guidelines for working with estimated limits:
+- Adding MaxItems, MaxProperties and MaxLength limits on all data accessed by CEL rules is the best practice.
+- O(n) - For simple rules, you can sometimes get away with iterating across a single map/list/string without a limit, but adding limits on all data accessed by CEL rules is the best practice
+- O(n^2)+ the product of the max lengths usually needs to be <1,000,000.  E.g. 1000 for 2 levels of nesting, 100 for 3 levels of nesting
+- O(n^3) - should generally be avoided
+
+Remember the big-O of a rule takes into account how many times the rule will be executed in addition to
```

Comment:
Directly start on explaining big-O might be confusing to the end user. I would suggest to either add more background on it or hide the details from end user. Might be worth to address what user should expect about the limit of their object and how they should be avoiding exceeding limit.